nexus-sdk/src/main/java/io/nexusrpc/handler/HandlerException.java

```diff
@@ -4,53 +4,124 @@
 
 /** Thrown from a handler for any unexpected error. */
 public class HandlerException extends RuntimeException {
+
+  /**
+   * RetryBehavior allows handlers to explicitly set the retry behavior of a {@link
+   * HandlerException}. If not specified, retry behavior is determined from the error type. For
+   * example {@link HandlerException.ErrorType#INTERNAL} is not retryable by default unless
+   * specified otherwise.
+   */
+  public enum RetryBehavior {
+    /**
+     * Indicates the retry behavior for a {@link HandlerException} is determined by the {@link
+     * HandlerException.ErrorType}.
+     */
+    UNSPECIFIED,
+    /**
+     * Indicates that a {@link HandlerException} should be retried, overriding the default retry
+     * behavior of the {@link HandlerException.ErrorType}.
+     */
+    RETRYABLE,
+    /**
+     * Indicates that a {@link HandlerException} should not be retried, overriding the default retry
+     * behavior of the {@link HandlerException.ErrorType}.
+     */
+    NON_RETRYABLE
+  }
+
   private final ErrorType errorType;
+  private final RetryBehavior retryBehavior;
 
   public HandlerException(ErrorType errorType, String message) {
-    this(errorType, new RuntimeException(message));
+    this(errorType, new RuntimeException(message), RetryBehavior.UNSPECIFIED);
+  }
+
+  public HandlerException(ErrorType errorType, String message, RetryBehavior retryBehavior) {
+    this(errorType, new RuntimeException(message), retryBehavior);
   }
 
   public HandlerException(ErrorType errorType, @Nullable Throwable cause) {
+    this(errorType, cause, RetryBehavior.UNSPECIFIED);
+  }
+
+  public HandlerException(
+      ErrorType errorType, @Nullable Throwable cause, RetryBehavior retryBehavior) {
     super(cause == null ? "handler error" : "handler error: " + cause.getMessage(), cause);
     this.errorType = errorType;
+    this.retryBehavior = retryBehavior;
   }
 
   /** Error type for this exception. */
   public ErrorType getErrorType() {
     return errorType;
   }
 
+  /** Retry behavior for this exception. */
+  public RetryBehavior getRetryBehavior() {
+    return retryBehavior;
+  }
+
+  public boolean isRetryable() {
+    if (retryBehavior != RetryBehavior.UNSPECIFIED) {
+      return retryBehavior == RetryBehavior.RETRYABLE;
+    }
+    switch (errorType) {
+      case BAD_REQUEST:
+      case UNAUTHENTICATED:
+      case UNAUTHORIZED:
+      case NOT_FOUND:
+      case NOT_IMPLEMENTED:
+        return false;
+      case RESOURCE_EXHAUSTED:
+      case INTERNAL:
+      case UNAVAILABLE:
+      case UPSTREAM_TIMEOUT:
+        return true;
+      default:
+        throw new IllegalStateException("Unknown error type: " + errorType);
```

Comment:
In Go we default to `true` just in case we later decide to add more error types and an old client is working with a newer server.

Reply:
Returning true here won't help with that scenario. The issue will first arise when deserializing the HandlerError/HandlerException. So in Go what error type would an unknown error type be deserialized to?

Reply:
In Go it's just a type alias to a string so you're not limited by the known set of enum values.

You can maybe add an UNKNOWN type or map it to INTERNAL.

Reply:
I see, I think most other SDKs will be like Java here and have a fixed set of enums. I am fine with UNKNOWN, but I would like this to be part of the of the Nexus API, any objection to me adding it there?